pubsub.go

```diff
@@ -133,6 +134,37 @@ type PubSubRouter interface {
 	Leave(topic string)
 }
 
+// PubSubControl is the control interface presented by the pubsub system to routers.
+type PubSubControl interface {
+	// PeersForTopic returns the set of peers in a topic; the boolean indicator will be false
+	// if the topic is unknown.
+	// This method is not safe to call outside the pubsub process loop.
+	PeersForTopic(topic string) (map[peer.ID]struct{}, bool)
```

Comment:
Hm, would a router want to limit their view of topic membership by only seeing peers that support their routing protocol? The reason I ask is because all other methods can be regarded as _events_, except for this one, which is pure state access. It seems to break the abstraction a bit.

I'd be more inclined to have the router and control layers be event-based, with no shared state. That could remove the need for the synthetic `Eval()` method -- which is only needed to protect state at an adjacent layer...

I see this model working as follows:

1. On startup, routers enrol with the control layer, providing a predicate that, given a `[]protocol.ID`, returns whether the router can interact with that peer or not.
2. On every new peer that advertises the umbrella pubsub protocol, we call the predicate on all routers, and emit a `NewPeer` event to routers that signal support.
3. The control layer tracks which routers are handling which peers. When a peer goes away, we emit an event relevant routers.
4. For topic membership control messages (join or leave), the control layer emit events to the supporting routers.
5. Routers send messages by pushing events on a channel that the control layer is looping over.
6. ...

This approach would insulate the state of each layer and we'd remove the need for cross-boundary scope changes.

Reply:
this is a major redesign i'd rather not tackle in this pr!

Reply:
Also, programming this with events makes it a lot harder to reason about; I really like the simplicity of a lock-free event loop.

Reply:
Hmmm, I get that, but pubsub/gossipsub is relied on by major downstream adopters, and I'm happy to invest time and effort in that refactor. I think it would make the code more decoupled, elegant, accessible for contributors, and less spaghetti.

Isn't the lock-free event loop inherently mono-threaded, anyway? It can become a major bottleneck.

If instead we let routers keep their own state locally, and be responsible for mutating in reaction to events they receive from the control layer, I think it makes the system more functional and a lot easier to reason about.

Reply:
👍 that having an `Eval()` function on the interface seems ridiculous (also, this being ridiculous was the reason given for #171 not being it's own router). If there's a short term need for users that don't want to fork go-libp2p-pubsub just to get their work done I'd rather be even grosser and just export the fields in PubSub so people know how unstable/temporary the state is.

A more approachable, and less intrusive, approach might be to invert a small amount of the design. For example `PeersForTopic` could easily be part of the `PubSubRouter` interface instead utilizing a `PubSubControl` interface. If certain structs/functions are really the same for various different routers then we could have some `PubSubRouterBase` struct that could be embedded in the individual routers.